contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -742,51 +768,152 @@ have a built-in `nil` value.
- 1. enables updates from users to be merged with changes made during late
-initialization, using strategic merge patch, as opposed to clobbering the
-change.
- 1. allow the component which does the late-initialization to use strategic
-merge patch, which facilitates composition and concurrency of such components.
-
-Although the apiserver Admission Control stage acts prior to object creation,
-Admission Control plugins should follow the Late Initialization conventions
-too, to allow their implementation to be later moved to a 'controller', or to
-client libraries.
+   (`patchStrategy:"merge"` attribute in the type definition).
+
+In particular we never want to change or override a value that was provided by
+the user.  If they requested something invalid, they should get an error.
+
+This:
+ 1. allows a user (with sufficient privilege) to override any system-default
+    behaviors by explicitly setting the fields that would otherwise have been
+    defaulted.
+ 1. enables updates from users to be merged with default values.
+
+### Considerations For PUT Operations
+
+Once an object has been created and defaults have been applied, it's very
+common for updates to happen over time.  Kubernetes offers several ways of
+updating an object which preserve existing values in fields other than those
+being updated (e.g. strategic merge patch and server-side apply).  There is,
+however, a less obvious way of updating objects which can have bad interactions
+with default values - PUT (aka `kubectl replace`).  While we recommend a
+read-modify-write cycle for PUTs, we know that not everyone does this and that
+even that is not a perfect solution in the face of version skew.
+
+When an object is updated with a PUT, the API server will see the "old"
+object with defaulted values and the "new" object without.  For static defaults
+this is only a problem if the CREATE and the PUT used different API versions,
+and even then it might actually be semantically correct.  For example, "v1" of
+an API might default a field to `false`, while "v2" defaults it to `true`.  If
+an object was created via API v1 (field = `false`) and then replaced via API
+v2, the field will attempt to change to `true`.
+
+For some APIs this is acceptable and actionable.  For others, this may be
+disallowed by validation.  In the latter case, the user will get an error about
+an attempt to change a field which is not even present in their YAML.  This is
+especially dangerous when adding new fields - an older client may not even know
+about the existence of the field, making even a read-modify-write cycle fail.
+While it is "correct" (in the sense that it is really what they asked for with
+PUT), it is not helpful and is a bad UX.
+
+When adding a field with a static or admission controlled default, this must be
+considered.  If the field is immutable after creation, consider adding logic to
+manually "patch" the value from the "old" object into the "new" one when it has
```

Comment:
for my understanding, you are saying that if my v1 immutable field defaults to false, and my v2 default to true; I have to mutate the field if the PUT doesn't specify a value on that field?

Reply:
The PUT with v2 will default the unspecified field to the v2 default value.  The validate-update code will see the field change from false (v1) to true (v2) and SHOULD fail with an error because the field is immutable.

The only fixes here are:

1) require RMW
2) don't use defaults.go, but do it like Service IPs - in registry - and manually patch it